### ISSUE TITLE
BMCLossMD loss Multi-dimensional

### ISSUE THREAD

**Issue by lijain** [jiawei-ren/BalancedMSE#10]
Thank you for sharing, I refer [BalancedMSE](https://github.com/jiawei-ren/BalancedMSE)/synthetic_benchmark/loss.py;
Modified loss function in distributed run regression, after training the result is very low is my code modified by mistake?
Below is the code for my change, which can explain why
![image](https://user-images.githubusercontent.com/18046568/170186355-6cb371b7-e632-4e71-9a36-ff7d0bff3a8d.png)

**Comment by jiawei-ren** [comment]
I can't tell how `sample_loc_mask` is computed from your screenshot, so I assume it to be some random sampling in the batch. Besides that, your code looks correct to me.

Here are some suggestions:
1. Make sure that your test set is balanced or you are using a balanced evaluation metric. Otherwise, Balanced MSE is not suitable for the task.
2. Disable the `noise_sigma` learning, i.e., do not add `noise_sigma` to the optimizer. Manually tune the `noise_sigma` parameter by gradually increasing it from a small value. When `noise_sigma` is small, the result should be identical to the standard MSE.

**Comment by lijain** [comment]
Thanks for the immediate answer：
    1、 I would like to ask what is the balance assessment indicator you mentioned above?
    2 、In addition, the purpose of my application is to solve some sample imbalance in the sample, the previous loss regression used a loss function like l1smooth, etc., using map evaluation criteria. The use of balancing mse is to slow down the problem of data this piece, this and MSE can not be directly replaced?
    3、 In the end, I didn't use learnable noise in this piece, just that

**Comment by jiawei-ren** [comment]
1. Balanced test set: the labels are uniformly distributed in the test set and there is no imbalance. Balanced evaluation metric: performance is averaged on all uniformly divided ranges, e.g., in age estimation, compute the error within each age and then average the error on all ages.
2. Smooth L1 version BMC is as follows. You may try the code. How is your mAP computed?

```python
def bmc_loss_smooth_l1_md(pred, target, noise_scale, beta=1.0):
    """Multi-dim Balanced Smooth-L1 Loss (BMC)
    Args:
      pred: A float tensor of size [batch, d].
      target: A float tensor of size [batch, d].
      noise_scale: A float number or tensor.
      beta: A float number. Hyper-parameter for Smooth L1 Loss
    Returns:
      loss: A float tensor. Balanced Smooth-L1 Loss.
    """
    logits = - F.smooth_l1_loss(pred.unsqueeze(1), target.unsqueeze(0), beta=beta, reduction='none') / noise_scale   # logit size: [batch, batch]
    logits = logits.mean(-1)
    loss = F.cross_entropy(logits, torch.arange(pred.shape[0]))     # contrastive-like loss
    loss = loss * noise_scale.detach()  # optional: restore the loss scale, 'detach' when noise is learnable

    return loss
```

**Comment by lijain** [comment]
Thank you for your immediate answers.Is that I applied the wrong version of the last version? So what scenario is it applied to? My MAP calculation is the coco kind

**Comment by jiawei-ren** [comment]
> Is that I applied the wrong version of the last version?  So what scenario is it applied to?

Both are correct. Using MSE or Smooth-L1 depends on your application. For bounding box regression, Smooth L1 should be the better option.

> My mAP calculation is the COCO kind.

I have checked the "spatial imbalance" in this [survey](https://arxiv.org/pdf/1909.00169.pdf), which is defined on the loss magnitude (i.e., regression loss tends to be overwhelmed by outliers) and is quite different from the setting in our paper. I would suggest first counting the offsets of positive anchor boxes to see if there is really an imbalance in regression targets. Moreover, mAP does not explicitly favor balanced box offset predictions. Therefore, I am not sure if Balanced MSE is applicable to your task.

**Comment by lijain** [comment]
Okay, thanks for the answer, I'm trying it out

**Comment by PaCaixiao** [comment]
May I ask if there is a derivation process for Smooth L1 version BMC

**Comment by jiawei-ren** [comment]
Hi,

I did not carefully derive it, the code above simply replaced MSE with smooth l1. It should be roughly correct since we are only changing the metric when computing the logits in BMC.
